File: api/reactor/src/main/java/com/coditory/sherlock/reactor/DelegatingDistributedLock.java
package com.coditory.sherlock.reactor;

import com.coditory.sherlock.LockRequest;
import com.coditory.sherlock.connector.AcquireResult;
import com.coditory.sherlock.connector.LockResultLogger;
import com.coditory.sherlock.connector.ReleaseResult;
import org.jetbrains.annotations.NotNull;
import reactor.core.publisher.Mono;

import java.time.Duration;

import static com.coditory.sherlock.Preconditions.expectNonEmpty;
import static com.coditory.sherlock.Preconditions.expectNonNull;

class DelegatingDistributedLock implements DistributedLock {
    private final LockResultLogger lockResultLogger;
    private final String lockId;
    private final String ownerId;
    private final Duration duration;
    private final AcquireAction acquireAction;
    private final ReleaseAction releaseAction;

    DelegatingDistributedLock(
        AcquireAction acquireAction,
        ReleaseAction releaseAction,
        String lockId,
        String ownerId,
        Duration duration
    ) {
        this.lockId = expectNonEmpty(lockId, "lockId");
        this.ownerId = expectNonEmpty(ownerId, "ownerId");
        this.duration = duration;
        this.acquireAction = expectNonNull(acquireAction, "acquireAction");
        this.releaseAction = expectNonNull(releaseAction, "releaseAction");
        this.lockResultLogger = new LockResultLogger(lockId, getClass());
    }

    @Override
    public String toString() {
        return "DelegatingDistributedLock{" +
            "lockId=" + lockId +
            ", ownerId=" + ownerId +
            ", duration=" + duration +
            '}';
    }

    @Override
    @NotNull
    public String getId() {
        return lockId;
    }

    @Override
    @NotNull
    public Mono<AcquireResult> acquire() {
        return acquire(new LockRequest(lockId, ownerId, duration));
    }

    @Override
    @NotNull
    public Mono<AcquireResult> acquire(@NotNull Duration duration) {
        expectNonNull(duration, "duration");
        return acquire(new LockRequest(lockId, ownerId, duration));
    }

    @Override
    @NotNull
    public Mono<AcquireResult> acquireForever() {
        return acquire(new LockRequest(lockId, ownerId, null));
    }

    @Override
    @NotNull
    public Mono<ReleaseResult> release() {
        return releaseAction.release(lockId, ownerId)
            .doOnNext(lockResultLogger::logResult);
    }

    private Mono<AcquireResult> acquire(LockRequest lockRequest) {
        expectNonNull(lockRequest, "lockRequest");
        return acquireAction.acquire(lockRequest)
            .doOnNext(lockResultLogger::logResult);
    }

    @FunctionalInterface
    public interface ReleaseAction {
        @NotNull
        Mono<ReleaseResult> release(@NotNull String lockId, @NotNull String ownerId);
    }

    @FunctionalInterface
    public interface AcquireAction {
        @NotNull
        Mono<AcquireResult> acquire(@NotNull LockRequest lockRequest);
    }
}

File: api/coroutine-connector/src/main/kotlin/com/coditory/sherlock/coroutines/DelegatingDistributedLock.kt
package com.coditory.sherlock.coroutines

import com.coditory.sherlock.LockDuration
import com.coditory.sherlock.LockId
import com.coditory.sherlock.LockRequest
import com.coditory.sherlock.OwnerId
import com.coditory.sherlock.Preconditions
import org.slf4j.LoggerFactory
import java.time.Duration

class DelegatingDistributedLock(
    private val acquireAction: AcquireAction,
    private val releaseAction: ReleaseAction,
    private val lockId: LockId,
    private val ownerId: OwnerId,
    private val duration: LockDuration,
) : DistributedLock {
    private val logger = LoggerFactory.getLogger(javaClass)
    override val id: String = lockId.value

    override suspend fun acquire(): Boolean {
        return acquire(LockRequest(lockId, ownerId, duration))
    }

    override suspend fun acquire(duration: Duration): Boolean {
        Preconditions.expectNonNull(duration, "duration")
        val lockDuration = LockDuration.of(duration)
        return acquire(LockRequest(lockId, ownerId, lockDuration))
    }

    override suspend fun acquireForever(): Boolean {
        return acquire(LockRequest(lockId, ownerId, null))
    }

    override suspend fun release(): Boolean {
        val released = releaseAction.release(lockId, ownerId)
        if (released) {
            logger.debug("Lock released: {}", lockId)
        } else {
            logger.debug("Lock not released: {}", lockId)
        }
        return released
    }

    private suspend fun acquire(lockRequest: LockRequest): Boolean {
        val acquired = acquireAction.acquire(lockRequest)
        if (acquired) {
            logger.debug("Lock acquired: {}, {}", lockId, lockRequest)
        } else {
            logger.debug("Lock not acquired: {}, {}", lockId, lockRequest)
        }
        return acquired
    }

    fun interface ReleaseAction {
        suspend fun release(
            lockId: LockId,
            ownerId: OwnerId,
        ): Boolean
    }

    fun interface AcquireAction {
        suspend fun acquire(lockRequest: LockRequest): Boolean
    }
}
